streamparse/storm/bolt.py

```diff
@@ -392,50 +385,14 @@ def emit_many(self, tups, **kwargs):
         kwargs['need_task_ids'] = False
         return super(BatchingBolt, self).emit_many(tups, **kwargs)
 
-    def _run(self):
-        """The inside of ``run``'s infinite loop.
-
-        Separated out so it can be properly unit tested.
-        """
-        tup = None
-        try:
-            tup = self.read_tuple()
-            if tup.task == -1 and tup.stream == '__heartbeat':
-                self.send_message({'command': 'sync'})
-            elif tup.component == '__system' and tup.stream == '__tick':
-                frequency = tup.values[0]
-                self.process_tick(frequency)
-            else:
-                group_key = self.group_key(tup)
-                with self._batch_lock:
-                    self._batches[group_key].append(tup)
-        except Exception as e:
-            log.error("Exception in %s.run() while adding %r to batch",
-                      self.__class__.__name__, tup, exc_info=True)
-            self.raise_exception(e)
-
-    def run(self):
-        """Modified and simplified run loop which runs in the main thread since
-        we only need to add tuples to the proper batch for later processing
-        in the _batcher thread.
-        """
-        storm_conf, context = self.read_handshake()
-        self._setup_component(storm_conf, context)
-
-        tup = None
-        self.initialize(storm_conf, context)
-        while True:
-            self._run()
+    def process_tick(self, tick_tup):
+        """Increment tick counter, and call ``process_batch`` for all current
+        batches if tick counter exceeds ``ticks_between_batches``.
 
-    def _batch_entry_run(self):
-        """The inside of ``_batch_entry``'s infinite loop.
-
-        Separated out so it can be properly unit tested.
+        See :class:`streamparse.storm.component.Bolt` for more information.
         """
-        time.sleep(self.secs_between_batches)
-        with self._batch_lock:
-            if not self._batches:
-                return # no tuples to save
+        self._tick_counter += 1
+        if self._tick_counter > self.ticks_between_batches and self._batches:
             for key, batch in iteritems(self._batches):
```

Comment:
Based on [this](https://github.com/Parsely/streamparse/blob/feature/simplify_batching_bolt/streamparse/storm/bolt.py#L442-L444) it's going to iterate over all batches and ignore `self.current_tups`.  However, even if it used `self.current_tups`, all unprocessed batches need to be failed, not just the one in progress. We need a way of  knowing which have been processed and which are awaiting processing.

Does Python3 no longer have the difference between `items` and `iteritems`?  With `items` you could pop because it freezes the set of `(key, values)` tuples before iterating.  If that's not an option you'll have to `keys = list(keys(self._batches))` or something and pop as you go through that list.

I don't think this is a new bug with your implementation, but it's definitely a bug.